### PR TITLE
Layer Effects: Support deleting layer effects

### PR DESCRIPTION
**Pull request by shaoshing**
(This PR requires https://github.com/adobe-photoshop/spaces-adapter/pull/120 to work)

### CODE REVIEW

**Comment by volfied** [comment]
I'll start reviewing this, while I'm doing that, can you rebase your changes to a concise commit?
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

File: src/img/ico-layer-delete.svg [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+<!-- Generator: Adobe Illustrator 19.0.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
+<svg version="1.1" id="layer-delete" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px"
+	 y="0px" viewBox="0 0 612 792" style="enable-background:new 0 0 612 792;" xml:space="preserve">
+<style type="text/css">
+	.st0{fill:none;stroke:#000000;stroke-width:80;stroke-miterlimit:10;}
+</style>
+<g transform="translate(1 1)">
+	<path class="st0" d="M408.6,291.4L201.4,498.6 M201.3,291.3l207.3,207.3"/>
+</g>
+</svg>
```
Comment (volfied) [inline src/img/ico-layer-delete.svg]: If this is your addition, can you check with @placegraphichere to make sure you get the final icon?

**Comment by volfied** [comment]
1 - Create a rectangle, add a drop shadow (X=5)
2 - Create another rectangle
3 - Select both rectangles, add a drop shadow (Y = 20)
4 - Delete either drop shadow
Result: All the shadows get deleted
Expected: Just the shadow I chose should be deleted

File: src/js/actions/layereffects.js [hunk 2]
```diff
@@ -179,6 +179,39 @@ define(function (require, exports) {
     setShadowEnabled.writes = [locks.PS_DOC, locks.JS_DOC];
 
     /**
+     * Delete the selected Shadow for all selected layers
+     *
+     * @param {Document} document
+     * @param {Immutable.Iterable.<Layer>} layers list of layers to update
+     * @param {number} shadowIndex index of the Drop Shadow within the layer(s)
```
Comment (volfied) [inline src/js/actions/layereffects.js]: Is this index going to be same for all layers?

Comment (volfied) [inline src/js/actions/layereffects.js]: I see you `forEach` to calculate actual index list. Please refer to bug I added as a comment.

File: src/js/models/layer.js [hunk 3]
```diff
@@ -134,12 +134,12 @@ define(function (require, exports, module) {
         /**
          * @type {Immutable.List.<Shadow>}
          */
```
Comment (volfied) [inline src/js/models/layer.js]: Why the change from null to `Immutable.List()`? We have other Lists in this record that initialize as null.

File: src/js/models/layer.js [hunk 4]
```diff
@@ -343,8 +357,7 @@ define(function (require, exports, module) {
             mode: "passThrough",
             proportionalScaling: false,
             isArtboard: false,
-            isLinked: false,
-            hasLayerEffect: false
```
Comment (volfied) [inline src/js/models/layer.js]: Groups can have effects as well.. so I can reproduce some of the play errors you were getting by adding/removing effects from group layers.

File: src/js/models/layerstructure.js [hunk 5]
```diff
@@ -1328,8 +1328,80 @@ define(function (require, exports, module) {
...
+        var nextLayers = Immutable.Map(layerIDs.reduce(function (map, layerID, index) {
+            var layer = this.byID(layerID),
+                layerEffects = layer.getLayerEffectsByType(layerEffectType) || Immutable.List(),
+                _layerEffectIndex,
+                layerEffect,
+                nextLayerEffect,
+                newProps,
+                nextLayer;
+
+            _layerEffectIndex = Immutable.List.isList(layerEffectIndex) ?
```
Comment (volfied) [inline src/js/models/layerstructure.js]: I think this style of declaring the var name, and then defining it below is not clean. Can you consolidate these?

**Comment by volfied** [comment]
I like where these changes are going, but I still see some bugs, especially with group layers.

I'd like to also hear your reasoning for changing `hasLayerEffects` to `usedToHaveLayerEffects`. I understand the need to add the latter, but why remove the first one?

Back to you, @shaoshing

File: src/img/ico-layer-delete.svg (hunk 1 above)
Comment (shaoshing) [inline src/img/ico-layer-delete.svg]: Got it from @designedbyscience as a temp icon, since we are moving toward UI redesign.

File: src/js/models/layer.js (hunk 3 above)
Comment (shaoshing) [inline src/js/models/layer.js]: I think a default value should be the value that works so that it doesn't rely on the users (us) to set a valid default value. In this case, I found I have to search over the files to make sure I can safely assume the dropShadows/innerShadows attr is always a list. Any thoughts?

File: src/js/models/layerstructure.js (hunk 5 above)
Comment (shaoshing) [inline src/js/models/layerstructure.js]: These are already existed in the master, don't know why it is showing as new here. Do you want me to fix them here?

**Comment by shaoshing** [comment]
```
1 - Create a rectangle, add a drop shadow (X=5)
2 - Create another rectangle
3 - Select both rectangles, add a drop shadow (Y = 20)
4 - Delete either drop shadow
Result: All the shadows get deleted
Expected: Just the shadow I chose should be deleted
```
Now deletion works correctly for selected:
* single / multiple layers 
* single / multiple groups
* layer + group

```
I'd like to also hear your reasoning for changing hasLayerEffects to usedToHaveLayerEffects. I understand the need to add the latter, but why remove the first one?
```

I am using *usedToHaveLayerEffects* to determine whether to use *setLayerEffect* or *setExtendedLayerEffect*. *setLayerEffect* is required only when the object is newly created and has no effect at all. After created its first effect, it will always has the layerEffect property, even after its layer effects are all deleted (remember the hidden layer effect). I am trying to replace a new variable name that represents its purpose correctly.

File: src/js/models/layer.js (hunk 3 above)
Comment (iwehrman) [inline src/js/models/layer.js]: I'm fine with initializing immutable record members like this to empty lists. Immutable is smart enough to only ever create a single empty list object, so there should be no performance impact. I would be _more_ concerned with making such a change consistently throughout our models, and also ensuring that client code no longer performs unnecessary falsey checks.

Comment (shaoshing) [inline src/js/models/layer.js]: there are about 10 Immutable.List attributes defined with null default value after a quick search, shall I refactor them to keep consistency or wait until our major release?

Comment (iwehrman) [inline src/js/models/layer.js]: I don't think you need to wait until after the release, though it might be better to submit a separate PR for that particular cleanup to make the review easier.

Comment (shaoshing) [inline src/js/models/layer.js]: roger that

**Comment by shaoshing** [comment]
back to you, Barkin

**Comment by volfied** [comment]
Please ping me again after a rebase and addressing https://github.com/adobe-photoshop/spaces-design/pull/1729/files#r34593446

**Comment by volfied** [comment]
:+1: Works great
